torch/cuda/amp/autocast_mode.py

```diff
@@ -0,0 +1,189 @@
+import torch
+import functools
+import warnings
+import numpy as np
+from torch._six import container_abcs, string_classes
+
+
+class autocast(object):
+    r"""
+    Instances of :class:`autocast` serve as context managers or decorators that
+    allow regions of your script to run in mixed precision.
+
+    In autocast-enabled regions, the backend automatically chooses the precision
+    for GPU operations to improve performance while maintaining accuracy.
+
+    When entering an autocast-enabled region, Tensors may be any type.  It is not necessary or
+    recommended to call ``.half()`` on your model(s) or data to use autocasting.
```

Comment:
I don't want to say something so absolute.  Calling `.half()` on the model is a third axis of mixed precision that can be manipulated modularly alongside autocasting and scaling.  It's ok to call .half() on your models and inputs if you have an optimizer that can handle FP16 weight updates (say an optimizer that manages its own master weights, or a stochastic rounding optimizer).

Reply:
Then either describe under what circumstances you would want or not want to call .half(), or remove this comment altogether. As is, it provides 0 bits of information. "You can either call .half() or not call .half(), but we don't describe and you shouldn't care what happens in either case".

Reply:
Hmm in that case I agree with original option.  Right now the only people with access to these optimizers are people who know that (for them) it's safe to call .half().  The "circumstances" under which it's safe aren't relevant/practical to Pytorch users in general so there's no point digressing about them.  If we ever reach a state where such optimizers are publicly available and well-documented so that I have something to point to, I'll update.